aten/src/ATen/core/boxing/KernelFunction.h

```diff
@@ -209,6 +250,11 @@ class TORCH_API KernelFunction final {
   template<bool AllowLegacyTypes = false, class Lambda>
   static std::enable_if_t<!guts::is_stateless_lambda<std::decay_t<Lambda>>::value, KernelFunction> makeFromUnboxedLambda(Lambda&& lambda);
 
+  template<bool AllowLegacyTypes = false, class Lambda>
+  static std::enable_if_t<guts::is_stateless_lambda<std::decay_t<Lambda>>::value, KernelFunction> makeFromUnboxedLambda_withKeys(Lambda&& lambda);
+  template<bool AllowLegacyTypes = false, class Lambda>
+  static std::enable_if_t<!guts::is_stateless_lambda<std::decay_t<Lambda>>::value, KernelFunction> makeFromUnboxedLambda_withKeys(Lambda&& lambda);
+
```

Comment:
Regarding the duplication here, I'm wondering if you can get away with having "less" functionality supported for this variant. For example, the lambda support here is so that people can say `impl(..., []() { ... })` which is cool but... you can always define it as a standalone function too. Since `withKeys` is intended for performance optimization, it seems plausible that less porcelain is OK in this situation. I'm not exactly sure how much this is used, but something to think about.

Reply:
That's a good point. Giving someone the option of registering a function pointer vs. a lambda vs. a functor in this specific case seems pretty unnecessary. The only version of the `withKeys` API that I use in this PR is the compile-time function pointer case. So we could maybe do away with the `withKeys` version of the functor API too.

That will cut down on the number of duplicate functions scattered throughout `KernelFunction.h`, but it doesn't address the fact that these functions really shouldn't need to be duplicated- since all that they're doing is forwarding the information along to other functions. Do you think it would look any cleaner to try templating these functions on a boolean, similar to `AllowLegacyTypes`? Then again, agreeing not to support this perf optimization for lambdas and/or functors will cut down on most of the duplication anyway.